### PR TITLE
Configure pprof on controller-manager.

### PR DESCRIPTION
**Pull request by cheftako**
Generated file.
Adding cluster mode changes.
Updated version for cluster mode.
Skipping version check on dev mode controllers.
Make pprof support case insensitive.

### Change description

Allows you to enable pprof and use it as follows.

apiVersion: customize.core.cnrm.cloud.google.com/v1alpha1
kind: NamespacedControllerReconciler
metadata:
  name: cnrm-controller-manager 
  namespace: kcc-test
spec:
  pprof:
    support: all
    port: 1234

kubectl port-forward cnrm-controller-manager-<id> -n cnrm-system 1234

curl http://localhost:1234/debug/pprof/heap > heap.outheap.out

Fixes #

### Tests you have done

- [x] Run `make ready-pr` to ensure this PR is ready for review.
- [x] Perform necessary E2E testing for changed resources.

### CODE REVIEW

File: go.work [hunk 1]
```diff
@@ -3,8 +3,8 @@ go 1.22.5
 use (
 	.
 	./dev/tools/controllerbuilder
-	./experiments/kompanion
-	./experiments/kubectl-plan
+//	./experiments/kompanion
```
Comment (justinsb) [inline go.work:6]: Looks unrelated - split into another PR maybe?

File: operator/config/crd/bases/customize.core.cnrm.cloud.google.com_controllerreconcilers.yaml [hunk 2]
```diff
@@ -41,6 +41,20 @@ spec:
           spec:
             description: ControllerReconcilerSpec is the specification of ControllerReconciler.
             properties:
+              pprofConfig:
```
Comment (justinsb) [inline operator/config/crd/bases/customize.core.cnrm.cloud.google.com_controllerreconcilers.yaml:44]: Naming nits: it's all config, so I would suggest `pprofConfig` -> `pprof` or `profiling`

And then if we use `pprof`, `pprofPort` can just be `port`, and `pprofSupport` can just be `support` (or `supportedModes`, I originally thought pprofSupport was really `enabled`

If we do `profiling`, `pprofPort` and `pprofSupport` (or `pprofSupportedModes` or `pprofModes`) make sense, because then we don't have pprof duplicated (and we allow for other means of profiling in future)

Comment (justinsb) [reply operator/config/crd/bases/customize.core.cnrm.cloud.google.com_controllerreconcilers.yaml:44]: BTW I also wonder whether otel might be more useful than pprof.  pprof might be useful for tracking down memory usage, and might be useful for tracking down CPU usage, but I don't _think_ either of those are our biggest problems?  Not a blocker to adding pprof support first though :-)

File: operator/config/crd/bases/customize.core.cnrm.cloud.google.com_namespacedcontrollerreconcilers.yaml [hunk 3]
```diff
@@ -41,6 +41,20 @@ spec:
           spec:
             description: NamespacedControllerReconciler is the specification of NamespacedControllerReconciler.
             properties:
+              pprofConfig:
+                description: PprofConfigures the debug endpoint on the service.
```
Comment (justinsb) [inline operator/config/crd/bases/customize.core.cnrm.cloud.google.com_namespacedcontrollerreconcilers.yaml:45]: Also nit, should the comment should probably be "Configures the debug endpoint on the service" or "<fieldName> configures the debug endpoint on the service"

File: operator/pkg/controllers/utils.go [hunk 4]
```diff
@@ -586,3 +586,86 @@ func applyRateLimitToContainerArg(container map[string]interface{}, rateLimit *c
...
+		return fmt.Errorf("pprof config customization for %s is not supported. "+
+			"Supported controllers: %s",
+			targetControllerName, strings.Join(customizev1alpha1.SupportedPprofControllers, ", "))
+	}
+
+	for _, item := range m.Items {
+		if item.GroupVersionKind() != targetControllerGVK {
+			continue
+		}
+		if !strings.HasPrefix(item.GetName(), targetControllerName) {
```
Comment (justinsb) [inline operator/pkg/controllers/utils.go:617]: It doesn't matter, but ... why the prefix match?

File: operator/pkg/controllers/utils.go [hunk 5]
```diff
@@ -586,3 +586,86 @@ func applyRateLimitToContainerArg(container map[string]interface{}, rateLimit *c
...
+			continue
+		}
+		if err := item.MutateContainers(customizePprofConfigFn(targetContainerName, pprofConfig)); err != nil {
+			return err
+		}
+		break // we already found the matching controller, no need to keep looking.
```
Comment (justinsb) [inline operator/pkg/controllers/utils.go:623]: My 2c: the small savings in CPU are not worth breaking.  I think it's better to return nil here and making it an error if we don't find a match (i.e. return error at the end of the function)

Another option is to count the number of containers changed and return an error if there was not exactly one changed.

File: operator/pkg/controllers/utils.go [hunk 6]
```diff
@@ -586,3 +586,86 @@ func applyRateLimitToContainerArg(container map[string]interface{}, rateLimit *c
...
+
+func applyPprofConfigToContainerArg(container map[string]interface{}, pprofConfig *customizev1alpha1.PprofConfig) error {
+	if pprofConfig == nil {
+		return nil
+	}
+	origArgs, found, err := unstructured.NestedStringSlice(container, "args")
+	if err != nil {
+		return fmt.Errorf("error getting args in container: %w", err)
+	}
+	wantArgs := []string{}
```
Comment (justinsb) [inline operator/pkg/controllers/utils.go:649]: LGTM.  Somewhere we have a similar function to replace flags, but I don't know which codebase that is in!

**Comment by justinsb** [comment]
Looks good, my only real quibble was with the field naming

File: operator/config/crd/bases/customize.core.cnrm.cloud.google.com_controllerreconcilers.yaml (hunk 2 above)
Comment (cheftako) [reply operator/config/crd/bases/customize.core.cnrm.cloud.google.com_controllerreconcilers.yaml:44]: This is really last mile wiring. We already have pprof capability in the binaries themselves. We were just missing a way to turn it on. All for adding otel but this seemed like a quick win.

File: operator/pkg/apis/core/customize/v1alpha1/controllerreconciler_types.go [hunk 7]
```diff
@@ -41,6 +41,9 @@ type NamespacedControllerReconcilerSpec struct {
 	// If not specified, the default will be Token Bucket with qps 20, burst 30.
 	// +optional
 	RateLimit *RateLimit `json:"rateLimit,omitempty"`
+	// PprofConfigures the debug endpoint on the service.
+	// +optional
+	PprofConfig *PprofConfig `json:"pprofConfig,omitempty"`
```
Comment (acpana) [inline operator/pkg/apis/core/customize/v1alpha1/controllerreconciler_types.go:46]: you can probably argue YAGNI against this but how about we sub class the Pprof config under something like `debugConfig` and we can add more debugging controls like otel, etc.

File: operator/config/crd/bases/customize.core.cnrm.cloud.google.com_controllerreconcilers.yaml (hunk 2 above)
Comment (cheftako) [reply operator/config/crd/bases/customize.core.cnrm.cloud.google.com_controllerreconcilers.yaml:44]: Also this was inspired by the LRO resources where the controller went silent. Figured this would be useful for confirming that the controller was stack waiting on the LRO and not that something else was wrong.

File: operator/pkg/controllers/utils.go (hunk 4 above)
Comment (cheftako) [reply operator/pkg/controllers/utils.go:617]: I believe the manifest name here is for the Pod(s) and would be something like "cnrm-controller-manager-okwcikx5o4yw4ly-0". 
However that needs to match the name of the kind: NamespacedControllerReconciler which in this case is just "cnrm-controller-manager".
Might be better to use the cnrm.cloud.google.com/component annotation but I would rather make that a broad fix across reconcilers.

File: operator/pkg/controllers/utils.go (hunk 5 above)
Comment (cheftako) [reply operator/pkg/controllers/utils.go:623]: Going with 2 but also changing ApplyContainerRateLimit to match.

File: go.work (hunk 1 above)
Comment (cheftako) [reply go.work:6]: I think I was having a problem with the linter or make command.. Let me see what happens if I just drop this change.

Comment (cheftako) [reply go.work:6]: Fixed for now :D

File: operator/pkg/apis/core/customize/v1alpha1/controllerreconciler_types.go (hunk 7 above)
Comment (cheftako) [reply operator/pkg/apis/core/customize/v1alpha1/controllerreconciler_types.go]: My personal take is if we think we will have many of these (say more than 5) then this organization would make a lot of sense.
Right now we have 1 and we'll see if we feel the need for the second.
It its just 2 it won't hurt but I'm not convinced its worth the effort of changing.
I may come to regret that.

Comment (jingyih) [reply operator/pkg/apis/core/customize/v1alpha1/controllerreconciler_types.go]: Since this is an alpha API, we can always make improvements in the future if needed.

File: operator/config/crd/bases/customize.core.cnrm.cloud.google.com_controllerreconcilers.yaml [hunk 8]
```diff
@@ -41,6 +41,20 @@ spec:
           spec:
             description: ControllerReconcilerSpec is the specification of ControllerReconciler.
             properties:
+              pprof:
+                description: Configures the debug endpoint on the service.
+                properties:
+                  pprofPort:
```
Comment (yuwenma) [inline operator/config/crd/bases/customize.core.cnrm.cloud.google.com_controllerreconcilers.yaml:47]: nit: `pprof` --> `port`

**Review by yuwenma — COMMENTED** [review]
/lgtm
/approve
/hold 

hold in case you want to rename the pprof naem. Feel free to unhold

**Review by yuwenma — COMMENTED** [review]
/lgtm

File: operator/pkg/controllers/utils.go (hunk 4 above)
Comment (justinsb) [reply operator/pkg/controllers/utils.go:622]: Ah, makes sense, thank you!  It shouldn't have the `-0`, but I had forgotten about the "namespace suffix"

**Review by justinsb — APPROVED** [review]
